### ISSUE TITLE
45$: Disable variable lookups in the presense of eval

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#95]

**Comment by AngelJA** [comment]
Can you give an example of what you mean?

I thought you meant quick-lint would give `use of undeclared variable` below, but it doesn't do any of that in string literals.
```javascript
eval('x + 5')
```

Strager's description of the problem, for reference: https://clips.twitch.tv/TransparentSuperStingrayBloodTrail

**Comment by strager** [comment]
Here's an example:
```js
eval("var x = 42;")
console.log(x);
```
This program prints `42`. quick-lint-js says that `x` is undeclared, but the `eval` declares `x`.

When quick-lint-js sees `eval`, it should assume that any variable could be declared by `eval`. Therefore, quick-lint-js should not report any error or warning for this code.

**Comment by singalhimanshu** [comment]
Are we planning to create an `eval-scope` for this ?

**Comment by strager** [comment]
> Are we planning to create an `eval-scope` for this ?

I don't know how I'd implement it.

**Comment by strager** [comment]
Perhaps scopes can contain a `bool has_eval = false;` member. If this member is true, disable use-of-undeclared-variable checks, and maybe some other checks too, for variable references in that scope. Propagate `has_eval` within a function, but not across functions.

**Comment by singalhimanshu** [comment]
> for variable references in that scope

Can you give an example of that?

**Comment by singalhimanshu** [comment]
I claim this for-hire task. I expect payment after I complete this task. I will email the quick-lint-js team if I am assigned this task.

**Comment by singalhimanshu** [comment]
> Perhaps scopes can contain a `bool has_eval = false;` member.

What if there is no scope how do we handle that?

**Comment by singalhimanshu** [comment]
> You mean the module scope? Or after `visit_end_of_module`?

Yes, I meant the module scope. If there is an `eval` in module scope we would have to disable use-of-undeclared-variable checks in the whole module I guess.

I say we ditch this `has_eval` idea and do something similar to #94 . What are the issues with doing it using an `eval_scope`?

**Comment by strager** [comment]
> If there is an `eval` in module scope we would have to disable use-of-undeclared-variable checks in the whole module I guess.

Yes, this sound right.

> I say we ditch the `has_eval` idea and do something similar to #94 . What are the issues with doing it using an `eval_scope`?

I don't see how this could work. How does the parser know when to call `visit_end_eval_scope`?

> If we use the `has_eval` idea we would have to traverse the `used_variables` vector at the end of function scope and module scope that might reduce some performance I guess if the number of used variables are more.

Your suggestion sounds like *not* using `has_eval`, which is fine by me.

I realized a case where `has_eval` wouldn't work:
```js
console.log(eval);  // sets scope.has_eval=true
let eval = 42;      // but eval is actually this variable, not the built-in eval function!
```
so I think we need to implement your idea of variable searching.

We could fix the performance problem by adding a `token_type` field to identifiers/variables. This would help with `arguments` too (https://github.com/quick-lint/quick-lint-js/blob/3fc909068830e98717bb0cdd634790e73993a064/src/lint.cpp#L514).
